### ISSUE TITLE
False positive PossiblyUndefinedVariable for reference in do/while loop

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#561]
Observed: PossiblyUndefinedVariable - src/preg_replace_callback.php:8:14 - Possibly undefined variable $count, first seen on line 7    
Expected: No error (http://php.net/manual/en/function.preg-replace-callback.php)

Aside: The callback map marks the variable at that position as w_count, which may or may not be useful for a general solution: `'preg_replace_callback' => ['string|array|null', 'regex'=>'string|array', 'callback'=>'callable', 'subject'=>'string|array', 'limit='=>'int', '&w_count='=>'int'],`

```php
<?php

function ab(string $value) : int {
    do {
        $value = preg_replace_callback('/aa/', function(array $match) : string{
            return 'ab';
        }, $value, -1, $count);
    } while ($count);
    echo $value;
    return $count;
}
```

**Comment by muglug** [comment]
Reproduced without `preg_replace_callback`:
```php
function foo(int &$bar = 3) : int {
    $bar = rand(0, 5);
    return 5;
}
function ab() : int {
    $value = "a";
    do {
        // setting value to itself triggers second loop check
        $value = foo($count);
    } while ($count);
    return $count;
}
```

**Comment by muglug** [comment]
Thanks!
